server/src/main/java/org/elasticsearch/cluster/metadata/ResettableValue.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.cluster.metadata;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.xcontent.ToXContent;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+/**
+ * This class holds a value of type @{param T} that can be in one of 3 states:
+ * - It has a concrete value, or
+ * - It is missing, or
+ * - It is meant to reset any other when it is composed with it.
+ * It is mainly used in template composition to capture the case when the user wished to reset any previous values.
+ * @param <T>
+ */
+public class ResettableValue<T> {
+    private static final ResettableValue<?> RESET = new ResettableValue<>(true, null);
+    private static final ResettableValue<?> UNDEFINED = new ResettableValue<>(false, null);
+    private static final String SKIP_RESET_VALUES = "reset";
+    private static final Map<String, String> SKIP_RESET_VALUES_PARAMS = Map.of(SKIP_RESET_VALUES, "true");
```

Comment:
I hear you, I was trying to capture the default value. So, if you do not do something, we see them, but you want you can explicitly disable them. Let me think how I can phrase it so it implies that the default is that it will be enabled.

Reply:
Let me know what you think.

Reply:
I'm thinking more along the lines of like `SHOW_RESET_VALUES` with a default of `true`, then the `SKIP_RESET_VALUES_PARAMS` could just be `Map.of(SHOW_RESET_VALUES, "false")`. 

That feels like it scans easier. I really only bring this up because there are some old IndicesOptions arguments from before the refactoring that have this problem.